### ISSUE TITLE
Star difficulty is (sometimes*) calculated wrong when switching rulesets

### ISSUE THREAD

**Issue by LittleEndu** [ppy/osu#11158]
\* doesn't seem to happen for every ruleset combination, but is consistent if that specific combination is broken

**Screenshots or videos showing encountered issue:** 
![stardiff](https://user-images.githubusercontent.com/11320483/101991741-4b017300-3cb7-11eb-8a49-d7faff957499.gif)

**osu!lazer version:** 
2020.1212.0

**Comment by peppy** [comment]
Could you explain what is wrong in more detail? Or what you'd expect to see? The issue I can see is that a mania beatmap is potentially being calculated using the catch ruleset during the transition, is that what you are trying to point out?

**Comment by LittleEndu** [comment]
Yeah, that's the issue. I would expect the difficulty icons to not update their star rating after it's calculated or the panels to leave faster so the wrong calculation isn't shown.

**Comment by peppy** [comment]
So the reason for this happening is that `BeatmapDifficultyCache` (located in `OsuGameBase`) is binding to the global ruleset bindable, while the rest of song select is using the local decoupled bindable.

Not sure on the correct way to fix this. The easiest solution I can devise is removing the decoupled bindable, but I believe this is intentionally here to avoid rapid ruleset changes causing great overhead.

**Comment by bdach** [comment]
I'm not sure that the decoupled ruleset is all that's wrong with the current setup (or even the only thing wrong with it, to be honest). It's possible for the difficulty cache to attempt to request diffcalc for a ruleset-specific beatmap, but with the wrong ruleset:

![2021-02-28-151714_949x333_scrot](https://user-images.githubusercontent.com/20418176/109421565-31c7b300-79d8-11eb-9ac3-d6b13926c7a9.png)

I feel like if conversion is not possible for a beatmap (which it is in this case), then the beatmap's ruleset should always be used? Although then again, that's a bit ill-defined in this context, since in that case the incoming ruleset parameter would essentially be ignored...

**Comment by smoogipoo** [comment]
Perhaps it's possible to create a new `BeatmapDifficultyCache` in the context you want? A global game-wide one and a local one for song select. Along this, probably separate out the tracking part into a separate component that uses a `BeatmapDifficultyCache` internally.

**Comment by bdach** [comment]
That doesn't do much for the issue, though. To expand on the above, consider a case of a mania beatmap, for instance: a mania difficulty icon shouldn't request a global-bindable-tracking difficulty anyway, since it doesn't make sense to run diffcalc for it in any ruleset other than mania.

I think the other issue that compounds the decoupled ruleset bindable stuff is that working beatmap changes happen just a little bit later than ruleset changes due to filter logic, so it's possible to have:

```
mania working beatmap, mania ruleset
mania working beatmap, catch ruleset
catch working beatmap, catch ruleset
```

**Comment by smoogipoo** [comment]
I think  it probably makes sense to use the beatmap's default ruleset in that case.
